enclave-runtime/src/sidechain_impl.rs

```diff
@@ -0,0 +1,268 @@
+	exec_tops, prepare_and_send_xts_and_block,
+	rpc::author::{alloc::sync::Arc, AuthorApi, OnBlockCreated, SendState},
+	state::load,
+	Result as EnclaveResult,
+};
+use codec::Encode;
+use core::time::Duration;
+use itc_light_client::{BlockNumberOps, LightClientState, NumberFor, Validator};
+use itp_ocall_api::EnclaveOnChainOCallApi;
+use itp_settings::sidechain::SLOT_DURATION;
+use itp_sgx_crypto::{Aes, AesSeal};
+use itp_sgx_io::SealedIO;
+use itp_storage_verifier::GetStorageVerified;
+use its_sidechain::{
+	aura::{Aura, AuraVerifier, SlotClaimStrategy},
+	consensus_common::{BlockImport, Environment, Error as ConsensusError, Proposal, Proposer},
+	primitives::traits::{Block as SidechainBlockT, ShardIdentifierFor, SignedBlock},
+	slots::{PerShardSlotWorkerScheduler, SlotInfo},
+	state::SidechainDB,
+	validateer_fetch::ValidateerFetch,
+};
+use log::error;
+use primitive_types::H256;
+use sgx_externalities::SgxExternalities;
+use sp_core::Pair;
+use sp_runtime::{traits::Block, MultiSignature};
+use sp_std::prelude::Vec;
+use std::{marker::PhantomData, string::ToString};
+
+///! `SlotProposer` instance that has access to everything needed to propose a sidechain block
+pub struct SlotProposer<PB: Block, SB: SignedBlock, Pair, OcallApi, LightClient, Author> {
+	pub ocall_api: Arc<OcallApi>,
+	pub light_client: Arc<LightClient>,
+	pub author: Arc<Author>,
+	pub proposer_key: Pair,
+	pub parentchain_header: PB::Header,
+	pub shard: ShardIdentifierFor<SB>,
+	_phantom: PhantomData<PB>,
+}
+
+///! `ProposerFactory` instance containing all the data to create the `SlotProposer` for the
+/// next `Slot`
+pub struct ProposerFactory<PB: Block, Pair, OcallApi, LightClient, Author> {
+	pub ocall_api: Arc<OcallApi>,
+	pub light_client: Arc<LightClient>,
+	pub author: Arc<Author>,
+	pub pair: Pair,
+	_phantom: PhantomData<PB>,
+}
+
+impl<PB: Block, Pair, OcallApi, LightClient, Author>
+	ProposerFactory<PB, Pair, OcallApi, LightClient, Author>
+{
+	pub fn new(
+		ocall_api: Arc<OcallApi>,
+		light_client: Arc<LightClient>,
+		author: Arc<Author>,
+		pair: Pair,
+	) -> Self {
+		Self { ocall_api, light_client, author, pair, _phantom: Default::default() }
+	}
+}
+
+impl<PB: Block<Hash = H256>, SB, P, OcallApi, LightClient, Author> Environment<PB, SB>
+	for ProposerFactory<PB, P, OcallApi, LightClient, Author>
+where
+	NumberFor<PB>: BlockNumberOps,
+	SB: SignedBlock<Public = sp_core::ed25519::Public, Signature = MultiSignature> + 'static,
+	SB::Block: SidechainBlockT<ShardIdentifier = H256, Public = sp_core::ed25519::Public>,
+	P: Pair,
+	P::Public: Encode,
+	OcallApi: EnclaveOnChainOCallApi + GetStorageVerified + 'static,
+	LightClient: Validator<PB> + Send + Sync + 'static,
+	Author: AuthorApi<H256, PB::Hash>
+		+ SendState<Hash = PB::Hash>
+		+ OnBlockCreated<Hash = PB::Hash>
+		+ Send
+		+ Sync,
+{
+	type Proposer = SlotProposer<PB, SB, P, OcallApi, LightClient, Author>;
+	type Error = ConsensusError;
+
+	fn init(
+		&mut self,
+		parent_header: &PB::Header,
```

Comment:
just a minor thing: why take it as reference, if you're cloning it anyway? shouldn't the caller be responsible for cloning, if necessary?

Reply:
You are right, this escaped me! 😄